src/main/java/io/cdap/plugin/salesforce/SObjectsDescribeResult.java

```diff
@@ -108,6 +134,92 @@ private void addSObjectDescribe(DescribeSObjectResult sObjectDescribe) {
 
     // sObjects names are case-insensitive
     // store them in lower case to ensure we obtain them case-insensitively
-    objectToFieldMap.put(sObjectDescribe.getName().toLowerCase(), fields);
+    objectToFieldMap.put(name.toLowerCase(), fields);
+  }
+
+  /**
+   * Describes last SObject by traversing though all parents names to the final SObject name.
+   * <p/>
+   * Input: [{"Account=TOP_LEVEL"}, {Account=TOP_LEVEL, Contacts=RELATIONSHIP}] <br/>
+   * Contacts is relationship name which real name is Contact.
+   * First we describe Account since its top level SObject,
+   * then based on Account describe result, we determine Contacts real name
+   * and describe Contacts by its real name Contact.
+   * Finally, we store Contact describe result under key named as Account.Contacts.
+   *
+   * @param connection Salesforce partner connection
+   * @param names list of parent SObject names and their types
+   * @param cache describe result cache
+   * @throws ConnectionException if unable to connect to Salesforce
+   */
+  private void addSObjectDescribe(PartnerConnection connection,
+                                  Map<String, SObjectDescriptor.SObjectType> names,
+                                  Map<String, DescribeSObjectResult> cache) throws ConnectionException {
+    DescribeSObjectResult previousDescribe = null;
+    for (Map.Entry<String, SObjectDescriptor.SObjectType> entry : names.entrySet()) {
+      String name = entry.getKey();
+      switch (entry.getValue()) {
+        case TOP_LEVEL:
+          previousDescribe = describe(connection, name, cache);
```

Comment:
It looks like this method requires that the map is an ordered map, which I believe is guaranteed with the creation of a LinkedHashMap somewhere else in the code, but could easily be broken at some point in the future.

I think this method would be less brittle and easier to understand if the top level object is passed in as one argument and all the collection of references and relationships are passed in as separate arguments. This would remove the need for this previousDescribe variable, remove the need for a switch on type, and remove the assumption about the map that is given.